src/bin/tree-sitter-graph/main.rs

```diff
@@ -60,13 +63,30 @@ fn main() -> Result<()> {
                 .long("allow-parse-errors")
                 .takes_value(false),
         )
+        .arg(
+            Arg::with_name("global")
+                .long("global")
+                .takes_value(true)
+                .action(ArgAction::Append),
+        )
         .get_matches();
 
     let tsg_path = Path::new(matches.value_of("tsg").unwrap());
     let source_path = Path::new(matches.value_of("source").unwrap());
     let current_dir = std::env::current_dir().unwrap();
     let quiet = matches.is_present("quiet");
     let lazy = matches.is_present("lazy");
+    let globals = matches.get_many::<String>("global").unwrap_or_default();
+    let mut globals_ = Variables::new();
+    for kv in globals {
+        let kv_ = kv
+            .split_once('=')
+            .with_context(|| format!("Expected key-value pair separated by '=', got {}.", kv))?;
+        globals_.add(
+            Identifier::from(kv_.0),
+            graph::Value::String(kv_.1.to_string()),
+        )?;
+    }
```

Comment:
I would have preferred to do this with a single map but I couldn't figure out how to express the two nested levels of failure (failure in parsing global values, and failure in the surrounding context). If anyone has suggestions, I'm all ears; I think we should be able to avoid both the second variable and the mutability, and (ideally) end up with something more readable as well.

Reply:
I actually like how you've done it here!  As you point out, it becomes messy with the two levels of error handling, and I think this rendering makes the control flow clearer.

(It's doable, and probably involves the [`collect` method](https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect), which takes a vec of results and turns it into a result of a vec.  But again, I'd rather we keep it as you've written it here.)

Reply:
(Also, if you want to bound the region where `globals_` is mutable, you can do something like this:

``` rust
let globals_ = globals_;
```

after you're done mutating it.  That makes it immutable for the remainder of the scope.)

Reply:
> I actually like how you've done it here! As you point out, it becomes messy with the two levels of error handling, and I think this rendering makes the control flow clearer.
> 
> (It's doable, and probably involves the [`collect` method](https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect), which takes a vec of results and turns it into a result of a vec. But again, I'd rather we keep it as you've written it here.)

Cool, if this is idiomatic then I'll be happy to :shipit:

> (Also, if you want to bound the region where `globals_` is mutable, you can do something like this:
> 
> ```rust
> let globals_ = globals_;
> ```
> 
> after you're done mutating it. That makes it immutable for the remainder of the scope.)

Oh, that's a lovely use of shadowing! Thank you.